contract/src/main/java/io/pravega/schemaregistry/contract/data/Compatibility.java

```diff
@@ -0,0 +1,203 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.contract.data;
+
+import io.pravega.common.ObjectBuilder;
+import lombok.Builder;
+import lombok.Data;
+
+/**
+ * Defines different Compatibility policy options for schema evolution for schemas within a group.
+ * The choice of compatibility policy tells the Schema Registry service whether a schema should be accepted to evolve
+ * into new schema by comparing it with one or more existing versions of the schema. 
+ * 
+ * {@link Type#AllowAny}: allow any changes to schema without any checks performed by the registry. 
+ * {@link Type#DenyAll}: disables any changes to the schema for the group.
+ * {@link Type#Backward}: a new schema can be used to read data written by previous schema. 
+ * {@link Type#BackwardTransitive}: a new schema can be used read data written by any of previous schemas. 
+ * {@link Type#BackwardTill}: a new schema can be used to read data written by any of previous schemas till schema 
+ * identified by version {@link Compatibility#backwardTill}. 
+ * {@link Type#Forward}: previous schema can be used to read data written by new schema. 
+ * {@link Type#ForwardTransitive}: all previous schemas can read data written by new schema. 
+ * {@link Type#ForwardTill}: All previous schemas till schema identified by version {@link Compatibility#forwardTill}
+ * can read data written by new schema. 
+ * {@link Type#Full}: both backward and forward compatibility.
+ * {@link Type#FullTransitive}: both backward and forward compatibility with all previous schemas.
+ * {@link Type#BackwardAndForwardTill}: All previous schemas till schema identified by version {@link Compatibility#forwardTill}
+ * can read data written by new schema. New schema can be used to read data written by any of previous schemas till schema 
+ * identified by version {@link Compatibility#backwardTill}. 
+ */
+@Data
+@Builder
+public class Compatibility implements SchemaValidationRule {
```

Comment:
```tkaitchuck 3 days ago Member
AllowAny, DenyAll, Full, and FullTransitive are just cases that are combinations of exisitng policies. IE:
FullTransitive = BackwardsTransitive + ForwardTransitive
If we instead had a separate backwards compatibility policy and forwards compatibility policy, there would only need to be at most 3 options for each and it could express even more policies than can be covered by these 10.
```
@shiveshr
shiveshr 5 days ago Author Collaborator
```
Yes, but that would make expressing the policy choice harder for the users. And understanding what a combination of values communicates makes it more complex to the user.
So common expressions are made available to user directly via the enum.
To express any complexity the BackwardAndForwardTill can be used to express any combination of backward and forward policy.
```
@tkaitchuck
tkaitchuck 3 days ago Member
```
    Yes, but that would make expressing the policy choice harder for the users.And understanding what a combination of values communicates makes it more complex to the user.

I don't agree that it is harder. It is necessary to understand the combinations now, it's just that they have different names which is more confusing because it requires yet more concepts to be learned. If there is only forward and backwards compatible, yes, that is two things that need to be explained and understood but 2 is better than 10+.

    To express any complexity the BackwardAndForwardTill can be used to express any combination of backward and forward policy.

Is this actually true? My understanding of Backwards was that it meant backwards for one version, which is not the same as BackwardsTill because BackwardsTill refers to a fixed point Backwards is a moving target. It is also not clear the BackwardAndForwardTill can include ForwardTransitive because new versions can be added which it's not obvious from the documentation how they would be included.

However if it is actually true that BackwardAndForwardTill based on the parameters can cover all the other options, then maybe we should use this as the only policy get rid of the enum and instead have better documentation/options for setting its parameters.

I think there is a way to do this effectivly. See my comment below.
```

@shiveshr
shiveshr 3 days ago Author Collaborator

```
You are absolutely right that our enum has redundancies and some of the rules can be collapsed into smaller set of rules.
For example AllowAny and BackwardTransitive and ForwardTransitive are redundant. They can be expressed using other rules:
BackwardTransitive = BackwardTill(v0)
ForwardTransitive = ForwardTill(v0)
AllowAny = dont specify anything.

Also- Full, Full transitive and BackwardTillAndForwardTill are actually composite rules that are pulled up as top level enum values.
So if allow for composite rules, we can get rid of all of these.

So above 6 rules can be reduced and we can get from 11 to 5 rules.

And with following 5 (Backward, Forward, BackwardTill and ForwardTill, DenyAll) we can express all others.

Now lets evaluate the benefit of reducing from 11 to 5 values -

    Lesser number of values that people will need to understand.
    counter point: Lesser number of values means they need to understand combinations to express their desired state.

    Can express more combinations than the 10 we have chosen to expose.
    If you look at it, only expressions missing from our enum that composite rules will allow are:
    BackwardOne + ForwardTill and ForwardOne+BackwardTill.
    All other combinations of composite rules can be reduced to backwardTill and forwardTill.

Using composite rules do not give any higher flexibility, but can lead to more complexity.
```

Reply:
You say:
> Using composite rules do not give any higher flexibility

But that is not true if:
>expressions missing from our enum that composite rules will allow are:
    BackwardOne + ForwardTill and ForwardOne+BackwardTill

So reducing the number to *is* more flexible. The argument "can lead to more complexity" is somewhat subjective, but I'd like to quantify complexity by looking at three things:
1. The number of conditional branches that need to exist in the code.
2. The number of code paths that lead to an invalid state.
3. The number of redundant code paths which must necessarily be identical. 
Let's score each of these approaches according to these three metrics:

For the 5 value approach:
1. There needs to be a branch to handle each case. But the combinations are additive and so don't create n^2 branches. - 5
2. Combining multiple forward or multiple backward policies or deny with others is invalid. - 3
3. There is no valid redundant way to specify anything - 0.

We could further improve the above by have a separate concept for a backwards compatibility policy and a forwards compatibility policy. Then the redundant policies could not even be represented.

For the 11 value approach:
1. There needs to be a branch to handle each case. And there needs to be a branch to check each existing value with the already applied value to see if they are compatible. - 110  (And yes I really would insist on having unit tests for every single one of these)
2. AllowAll and DenyAll can't be combined with anything else - 2. Plus for each of forward and backward there are 4 invalid combinations with each item, and three valid ones. - 12 and the ForwardAndBackwards can be combined with one and til but not in other ways, so - 2 for a total of 16.
3. Each of these items not in the smaller list can be expressed identically using the smaller items. - 6

Given that we need to write integration tests for each of these cases (and some are redundant), I consider the later approach much worse. It also runs afoul of "Make illegal states unrepresentable" best practice. (If you aren't familiar with this idea, this talk explains it starting at 18:00 https://vimeo.com/14313378)

Reply:
The complexity I speak of is in terms of users expressing their intent. 
With composite rules, a user could end up expressing rules of following kind:
`BackwardTill(v1) + BackwardTill(v2)` 
This is reducible to `BackwardTill(min(v1, v2))` but allowing for composite rules means anything could be expressed. 
There are two things that i dont like with this:
1. we will need to reduce the user supplied expression into a reduced form. The user says A and when they check back we show them A'. So they need to reconcile the view we created and the intent they wanted to express. 
2. It allows for mixing of two concepts --> example: backward-one and forward till. `till` expressions declare a check till a fixed point. While `-one` is a moving target. This is not particularly useful and will be hard for the user to comprehend the semantics of what the check for the new schema will be. 

By giving them a limited set of well defined expressions, it becomes easier to choose and express the intent. 
Let us first ignore the new policies that we have introduced that end with `till` suffix. 
If we look at other expressions, they are concepts that are fairly simple to understand and have been standardized by existing schema registries and accepted by users. 
These are namely - Backward, Forward, Full (each with transitive property). 

We are introducing additional constructs that give more power to the user by adding `BackwardTill` and `ForwardTill`. These two are also fairly easy to follow in this same context. 

The only complex concept to understand for the user becomes `BackwardTillAndForwardTill` which is effectively a composite type policy we give to the user to express complex intent should they wish to do so. It is for advanced users, and only for situations where users want to deal with the complexity of such a declaration. 


Now let us look at the code complexity argument. 
The code path complexity is not very different for 11 vs 5 because similar checks need to be executed for both expressions. 

From code path checks --> we have two logical functions:
1. canRead(schema1, schema2) ==> can schema1 be used to read data written using schema2 
2. canBeRead(schema1, schema2) ==> can data written with schema1 be read using schema2

All policies result in determining which schemas need to be checked against the new schema. 

The difference between `5` vs `11` is  switch case block.
for composite rules: 
loop rules {
switch ruleType: input: one or more older schemas and call canRead or canBeRead or both
}

For `11` it looks as follows:
switch ruleType: input: one or more older schemas and call canRead or canBeRead or both

Unit test coverage for correction will need to cover similar cases for either. 
In fact lesser code paths for `11` case as we dont actually allow for backwardOne+forwardTill and forwardOne+backwardTill compositions. 

> There needs to be a branch to handle each case. And there needs to be a branch to check each existing value with the already applied value to see if they are compatible. - 110 (And yes I really would insist on having unit tests for every single one of these)

I dont know how you arrived at the number of `110` there though. 

> Plus for each of forward and backward there are 4 invalid combinations with each item, and three valid ones. - 12 and the ForwardAndBackwards can be combined with one and til but not in other ways, so - 2 for a total of 16.

Again i dont follow this. Do note that we do not allow for multiple rules of type `Compatibility` to be specified. There is exactly one compatibility rule. Which has one of 11 options to choose from.
So there is no combination. 

There are 11 different unit tests and 11 code paths.

Reply:
Now let us look at the code complexity argument. 
The code path complexity is not very different for 11 vs 5 because similar checks need to be executed for both expressions. 

From code path checks --> we have two logical functions:
1. canRead(schema1, schema2) ==> can schema1 be used to read data written using schema2 
2. canBeRead(schema1, schema2) ==> can data written with schema1 be read using schema2

All policies result in determining which schemas need to be checked against the new schema. 

The difference between `5` vs `11` is  switch case block.
for composite rules: 
```
loop rules {
switch ruleType: {
// 5 cases.
// processing for a case: input: one or more older schemas and call canRead or canBeRead or both
}
}
```

For `11` it looks as follows:
```
// 11 cases.
// processing for a case: input: one or more older schemas and call canRead or canBeRead or both
```

Unit test coverage for correction will need to cover similar cases for either. 
In fact lesser code paths for `11` case as we dont actually allow for backwardOne+forwardTill and forwardOne+backwardTill compositions. 

> There needs to be a branch to handle each case. And there needs to be a branch to check each existing value with the already applied value to see if they are compatible. - 110 (And yes I really would insist on having unit tests for every single one of these)

I dont know how you arrived at the number of `110` there though. 

> Plus for each of forward and backward there are 4 invalid combinations with each item, and three valid ones. - 12 and the ForwardAndBackwards can be combined with one and til but not in other ways, so - 2 for a total of 16.

Again i dont follow this. Do note that we do not allow for multiple rules of type `Compatibility` to be specified. There is exactly one compatibility rule. Which has one of 11 options to choose from.
So there is no combination. 

There are 11 different unit tests and 11 code paths.

Reply:
> I dont know how you arrived at the number of 110 there though.

Nothing in the API right now stops users from specifying multiple SchemaValidationRules. It explicitly takes a list. The implication being that one can pass as many as one likes. This means we need to test all possible combinations of policies as they can be combined arbitrary. That is at least 10*11, but could even be more, as it is technically possible to set 3 or more policies, it's just that one of those is always redundant so I excluded that for simplicity. 

I am arguing not that 5 is better than 11, but rather than having a fixed SINGLE forwardCompatabilityRule and a single backwardCompatabilityRule is simpler than taking a list of rules which can be any combination or order of forward, backward, or composite policy. We have to accept the fact that any thing possible to express in the REST client WILL be sent by some application, because we are dealing with generated code which will not have the benefit of reading your documentation.

Reply:
> Nothing in the API right now stops users from specifying multiple SchemaValidationRules. It explicitly takes a list. The implication being that one can pass as many as one likes
The Rules are set as a Map<String, ValidationRule>. And only rule allowed for is Compatibility (both on rest DTO and in the model object). 

So firstly, there can be exactly one rule of type Compatibility. Secondly, service only accepts rules of type compatibiilty and rejects anything else as the only rule supported is compatibility. 

>I am arguing not that 5 is better than 11, but rather than having a fixed SINGLE forwardCompatabilityRule and a single backwardCompatabilityRule is simpler than taking a list of rules which can be any combination or order of forward, backward, or composite policy

Yes, then we are on the same page and what you want is expressed where user can choose exactly one of 11 choices we give. 

> I am also very concerned about the redundant combinations created. See my comments on 3 above.
There are no "redundant" combinations because there is no combination. Everything is at a flat level with one of 11 choices. If we were to allow for combinations, then we would not have required 11 choices and those would have been redundant. But we do not allow for combinations.

Reply:
Based on the offline discussion, i have removed the wrapping SchemaValidationRules class.
And remodeled Compatibility to break it into BackwardPolicy and ForwardPolicy.